src/os/bluestore/BlueStore.cc

```diff
@@ -1170,41 +1173,42 @@ struct LruOnodeCacheShard : public BlueStore::OnodeCacheShard {
     ceph_assert(p != lru.begin());
     --p;
     ceph_assert(num >= n);
-    num -= n;
     while (n-- > 0) {
```

Comment:
Here might be one place that is broken. 
The original _trim_to can really trim n onodes, but now it might fail to do so as some onode's pin_nref > 1. 
Not sure if there is a way to make up it here.

Reply:
Indeed the above points are valid. Not that the previous behavior was totally wrong - as we call trimming on every update op it would eventually do the proper trimming even if specific iteration doesn't fulfill the task completely...

Anyway your proposal results in a more elegant solution indeed. Please see the updated PR, cache trimming rework is Implemented as an individual commit.